src/codegen/compile.ml

```diff
@@ -2344,6 +2347,57 @@ module MakeCompact (Num : BigNumType) : BigNumType = struct
           get_res
       end)
 
+
+  let compile_lsh env =
+    Func.share_code2 env "B_lsh" (("n", I32Type), ("amount", I32Type)) [I32Type]
+    (fun env get_n get_amount ->
+      get_n ^^
+      BitTagged.if_tagged_scalar env [I32Type]
+        ( (* non-ptr, shift amount <= 42: signed i32 -> i64; shift left, remember, trunc to i32, extend, compare with recalled, if same, done! *)
+          get_n ^^
+          G.i (Convert (Wasm.Values.I64 I64Op.ExtendSI32)) ^^
+          get_amount ^^
+          G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
+          G.i (Binary (Wasm.Values.I64 I64Op.Shl)) ^^
+          let set_remember, get_remember = new_local64 env "remember" in
+          set_remember ^^ get_remember ^^
+          G.i (Convert (Wasm.Values.I32 I32Op.WrapI64)) ^^
+          let set_res, get_res = new_local env "res" in
+          set_res ^^ get_res ^^
+          G.i (Convert (Wasm.Values.I64 I64Op.ExtendSI32)) ^^ (* exclude sign flip *)
+          get_remember ^^
+          G.i (Compare (Wasm.Values.I64 I64Op.Eq)) ^^
+          get_amount ^^ compile_rel_const I32Op.LeU 42l ^^
+          G.i (Binary (Wasm.Values.I32 I32Op.And)) ^^
+          G.if1 I32Type
+            get_res
+            (get_n ^^ compile_shrS_const 1l ^^ Num.from_word30 env ^^ get_amount ^^ Num.compile_lsh env)
+        )
+        (get_n ^^ get_amount ^^ Num.compile_lsh env))
```

Comment:
I assume the Num operations do the compaction of result when necessary, right?

Reply:
No. Any proper bignum shifted to the left will remain a bignum. No compaction can happen here.

Reply:
> Doesn't it also make sense to have shifts on Ints?

I am not entirely sure about that yet. Bit manipulations presume a bit pattern representation and we never say what that is for arbitrary precision `Int`s. Internally it is two's complement for compact `Int`s, but TomMathLib uses a different format (1 bit sign + absolute mantissa). We _could_ define the shift operators as-if `Int` was `base * 2 ** exp`.

Reply:
Also if we have a shift operator for `Int`, somebody will appear and ask for rotation operators :-)

Reply:
Github is acting up again. The last two answers should have gone into the review comment thread. Oh well.